src/lib/io/sol-ipm.c

```diff
@@ -0,0 +1,323 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <errno.h>
+#include <inttypes.h>
+#include <stdlib.h>
+
+#ifndef SOL_LOG_DOMAIN
+#define SOL_LOG_DOMAIN &_sol_ipm_log_domain
+#endif
+
+#include "sol-log-internal.h"
+
+#include "ipm.h"
+#include "ipm/ipm_quark_se.h"
+
+#include "sol-log.h"
+#include "sol-macros.h"
+#include "sol-mainloop-zephyr.h"
+#include "sol-types.h"
+#include "sol-util.h"
+#include "sol-vector.h"
+
+SOL_LOG_INTERNAL_DECLARE_STATIC(_sol_ipm_log_domain, "ipm");
+
+#define SOL_IPM_ID_FREE_REMOTE 0
+
+/* Declare channels we are going to use.
+ * Channel 0 is Quark -> ARC
+ * Channel 1 is ARC -> Quark
+ */
+#ifdef CONFIG_X86
+QUARK_SE_IPM_DEFINE(message_ipm_receiver, 1, QUARK_SE_IPM_INBOUND);
+QUARK_SE_IPM_DEFINE(message_ipm_sender, 0, QUARK_SE_IPM_OUTBOUND);
+#elif CONFIG_ARC
+#warning "Building for ARC"
+QUARK_SE_IPM_DEFINE(message_ipm_receiver, 0, QUARK_SE_IPM_INBOUND);
+QUARK_SE_IPM_DEFINE(message_ipm_sender, 1, QUARK_SE_IPM_OUTBOUND);
+#endif
+
+/*XXX there MUST be a better way of doing so */
+#define AS_PTR(_ptr) *(uint64_t *)_ptr
+
+int sol_ipm_init(void);
+void sol_ipm_shutdown(void);
+
+static void free_ipm_blob(struct sol_blob *blob);
+
+struct sol_blob_ipm {
+    struct sol_blob base;
+    struct sol_blob *remote;
+};
+
+static const struct sol_blob_type SOL_BLOB_TYPE_IPM = {
+    SOL_SET_API_VERSION(.api_version = SOL_BLOB_TYPE_API_VERSION, )
+    SOL_SET_API_VERSION(.sub_api = 0, )
+    .free = free_ipm_blob,
+};
+
+struct sol_ipm_receiver_entry {
+    uint32_t id;
+    void (*receive_cb)(void *cb_data, uint32_t id, struct sol_blob *message);
+    const void *data;
+    struct sol_blob *last_remote_blob; /* Hackish... I store last blob received here, to avoid having yet another list to store them */
+};
+
+struct sol_ipm_sent_message {
+    uint32_t id;
+    struct sol_blob *message;
+    void (*message_consumed_cb)(void *cb_data, uint32_t id, struct sol_blob *message);
+    const void *data;
+};
+
+static struct sol_vector ipm_receivers = SOL_VECTOR_INIT(struct sol_ipm_receiver_entry);
+static struct sol_vector ipm_sent_messages = SOL_VECTOR_INIT(struct sol_ipm_sent_message);
+static struct device *ipm_sender, *ipm_receiver;
+static uint32_t ipm_max_id;
+
+static void
+free_ipm_blob(struct sol_blob *blob)
+{
+    struct sol_blob_ipm *ipm_blob = (struct sol_blob_ipm *)blob;
+    int r;
+
+    if (blob->type != &SOL_BLOB_TYPE_IPM) {
+        SOL_WRN("Invalid blob type: expected SOL_BLOB_TYPE_IPM(%p), got: %p",
+            &SOL_BLOB_TYPE_IPM, blob->type);
+        return;
+    }
+
+    r = ipm_send(ipm_sender, 1, SOL_IPM_ID_FREE_REMOTE, &ipm_blob->remote,
```

Comment:
Not sure if I got:
`ipm_blob->remote` has type `struct sol_blob *`. Let's say it's content is `0x40001234` - memory position of a `struct sol_blob`.
`ipm_send` takes a `void *` and a `size`, it will put what is pointed by, up to `size` bytes.
So, if send only `ipm_blob->remote`, it will look what is at position `0x40001234`, and get `size` (4) bytes there. But this is actually part of the struct itself.
That's why a send `&ipm_blob->remote`. This way, `0x40001234` will be on channel registers - and this value is then, on remote side, cast back to a pointer to `struct sol_blob`.